File: frontend/src/app/services/portfolio.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';
import {
  Profile,
  Experience,
  Project,
  SkillCategory,
  Education,
  Certification,
  ContactSettings
} from '../models';

/**
 * Portfolio service for fetching public portfolio data
 */
@Injectable({
  providedIn: 'root'
})
export class PortfolioService {
  private readonly api = inject(ApiService);

  /**
   * Get profile information
   */
  getProfile(): Observable<Profile> {
    return this.api.get<Profile>('/api/public/profile');
  }

  /**
   * Get work experience list
   */
  getExperiences(): Observable<Experience[]> {
    return this.api.get<Experience[]>('/api/public/experience');
  }

  /**
   * Get projects list
   */
  getProjects(): Observable<Project[]> {
    return this.api.get<Project[]>('/api/public/projects');
  }

  /**
   * Get featured projects
   */
  getFeaturedProjects(): Observable<Project[]> {
    return this.api.get<Project[]>('/api/public/projects?featured=true');
  }

  /**
   * Get skill categories with items
   */
  getSkills(): Observable<SkillCategory[]> {
    return this.api.get<SkillCategory[]>('/api/public/skills');
  }

  /**
   * Get education list
   */
  getEducation(): Observable<Education[]> {
    return this.api.get<Education[]>('/api/public/education');
  }

  /**
   * Get certifications list
   */
  getCertifications(): Observable<Certification[]> {
    return this.api.get<Certification[]>('/api/public/certifications');
  }

  /**
   * Get contact settings
   */
  getContactSettings(): Observable<ContactSettings> {
    return this.api.get<ContactSettings>('/api/public/contact');
  }

  /**
   * Check API health
   */
  checkHealth(): Observable<{ status: string; timestamp: string }> {
    return this.api.get<{ status: string; timestamp: string }>('/api/public/health');
  }
}

File: backend/src/main/java/com/sita/portfolio/service/PortfolioService.java
package com.sita.portfolio.service;

import com.sita.portfolio.model.dto.response.*;
import com.sita.portfolio.repository.*;
import com.sita.portfolio.service.mapper.EntityMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Service for retrieving published portfolio data.
 * All methods return only published content, sorted by sort_order.
 */
@Service
@RequiredArgsConstructor
@Slf4j
@Transactional(readOnly = true)
public class PortfolioService {

    private final ProfileRepository profileRepository;
    private final ExperienceRepository experienceRepository;
    private final ProjectRepository projectRepository;
    private final SkillCategoryRepository skillCategoryRepository;
    private final EducationRepository educationRepository;
    private final CertificationRepository certificationRepository;
    private final ContactSettingsRepository contactSettingsRepository;
    private final EntityMapper mapper;

    /**
     * Gets all published portfolio data in a single response.
     */
    public PortfolioResponse getPortfolio() {
        log.debug("Fetching complete portfolio data");

        return PortfolioResponse.builder()
                .profile(getProfile())
                .experiences(getPublishedExperiences())
                .projects(getPublishedProjects())
                .featuredProjects(getFeaturedProjects())
                .skills(getPublishedSkills())
                .education(getPublishedEducation())
                .certifications(getPublishedCertifications())
                .contact(getContactSettings())
                .build();
    }

    /**
     * Gets the profile.
     */
    public ProfileResponse getProfile() {
        return profileRepository.findProfile()
                .map(mapper::toProfileResponse)
                .orElse(null);
    }

    /**
     * Gets all published experiences sorted by sort_order.
     */
    public List<ExperienceResponse> getPublishedExperiences() {
        return mapper.toExperienceResponseList(
                experienceRepository.findByPublishedTrueOrderBySortOrderAsc()
        );
    }

    /**
     * Gets all published projects sorted by sort_order.
     */
    public List<ProjectResponse> getPublishedProjects() {
        return mapper.toProjectResponseList(
                projectRepository.findByPublishedTrueOrderBySortOrderAsc()
        );
    }

    /**
     * Gets all featured and published projects sorted by sort_order.
     */
    public List<ProjectResponse> getFeaturedProjects() {
        return mapper.toProjectResponseList(
                projectRepository.findByFeaturedTrueAndPublishedTrueOrderBySortOrderAsc()
        );
    }

    /**
     * Gets a published project by slug.
     */
    public ProjectResponse getProjectBySlug(String slug) {
        return projectRepository.findBySlugAndPublishedTrue(slug)
                .map(mapper::toProjectResponse)
                .orElse(null);
    }

    /**
     * Gets all published skill categories with their skills, sorted by sort_order.
     */
    public List<SkillCategoryResponse> getPublishedSkills() {
        return mapper.toSkillCategoryResponseList(
                skillCategoryRepository.findByPublishedTrueOrderBySortOrderAsc()
        );
    }

    /**
     * Gets all published education entries sorted by sort_order.
     */
    public List<EducationResponse> getPublishedEducation() {
        return mapper.toEducationResponseList(
                educationRepository.findByPublishedTrueOrderBySortOrderAsc()
        );
    }

    /**
     * Gets all published certifications sorted by sort_order.
     */
    public List<CertificationResponse> getPublishedCertifications() {
        return mapper.toCertificationResponseList(
                certificationRepository.findByPublishedTrueOrderBySortOrderAsc()
        );
    }

    /**
     * Gets the contact settings.
     */
    public ContactSettingsResponse getContactSettings() {
        return contactSettingsRepository.findSettings()
                .map(mapper::toContactSettingsResponse)
                .orElse(null);
    }

}
